lib/vnet/vnet_test.go

```diff
@@ -0,0 +1,281 @@
+	require.NoError(t, err)
+	testStack.SetRouteTable([]tcpip.Route{{
+		Destination: subnet,
+		NIC:         nicID,
+	}})
+
+	// Create the VNet and connect it to the other side of the TUN.
+	manager, err := NewManager(&Config{
+		TUNDevice:  tun2,
+		IPv6Prefix: vnetIPv6Prefix,
+	})
+	require.NoError(t, err)
+
+	utils.RunTestBackgroundTask(ctx, t, &utils.TestBackgroundTask{
+		Name: "VNet",
+		Task: func(ctx context.Context) error {
+			if err := manager.Run(ctx); !errIsOK(err) {
+				return trace.Wrap(err)
+			}
+			return nil
+		},
+		Terminate: func() error {
+			tunErr := tun2.Close()
+			destroyErr := manager.Destroy()
+			return trace.NewAggregate(tunErr, destroyErr)
+		},
+	})
+
+	return &testPack{
+		vnetIPv6Prefix:   vnetIPv6Prefix,
+		manager:          manager,
+		testStack:        testStack,
+		testLinkEndpoint: linkEndpoint,
+		localAddress:     localAddress,
+	}
+}
+
+// dial dials the VNet address [addr] from the test virtual netstack.
+func (p *testPack) dial(ctx context.Context, addr tcpip.Address) (*gonet.TCPConn, error) {
+	conn, err := gonet.DialTCPWithBind(
+		ctx,
+		p.testStack,
+		tcpip.FullAddress{
+			NIC:      nicID,
+			Addr:     p.localAddress,
+			LinkAddr: p.testLinkEndpoint.LinkAddress(),
+		},
+		tcpip.FullAddress{
+			NIC:      nicID,
+			Addr:     addr,
+			Port:     456,
+			LinkAddr: p.manager.linkEndpoint.LinkAddress(),
+		},
+		ipv6.ProtocolNumber,
+	)
+	return conn, trace.Wrap(err)
+}
+
+// TestVnetEcho is a preliminary VNet test that manually installs an echo handler on a specific IP, TCP dials
+// it, and makes sure writes are echoed back to the TCP conn.
+func TestVnetEcho(t *testing.T) {
+	ctx := context.Background()
+	p := newTestPack(t, ctx)
+
+	dialAddress, err := p.manager.assignTCPHandler(echoHandler{})
+	require.NoError(t, err)
+
+	for i := 0; i < 10; i++ {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			t.Parallel()
+
+			conn, err := p.dial(ctx, dialAddress)
+			require.NoError(t, err)
+			defer conn.Close()
+
+			testString := "Hello, World!\n"
+			_, err = conn.Write([]byte(testString))
+			require.NoError(t, err)
+			t.Cleanup(func() { require.NoError(t, conn.Close()) })
+
+			scanner := bufio.NewScanner(conn)
+			require.True(t, scanner.Scan(), "scan failed: %v", scanner.Err())
+			line := scanner.Text()
+			require.Equal(t, strings.TrimSuffix(testString, "\n"), line)
+		})
+	}
+}
+
+type echoHandler struct{}
+
+func (echoHandler) handleTCP(ctx context.Context, connector tcpConnector) error {
+	conn, err := connector()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close()
+	_, err = io.Copy(conn, conn)
+	return trace.Wrap(err)
+}
+
+func randomULAAddress() (tcpip.Address, error) {
+	prefix, err := IPv6Prefix()
+	if err != nil {
+		return tcpip.Address{}, trace.Wrap(err)
+	}
+	bytes := prefix.As16()
+	bytes[15] = 2
+	return tcpip.AddrFrom16(bytes), nil
+}
+
+var fakeTUNClosedErr = errors.New("TUN closed")
```

Comment:
My editor complains `error var fakeTUNClosedErr should have name of the form errFoo (ST1012)go-staticcheck`